cf/test/test_Data.py

```diff
@@ -131,6 +131,149 @@ class DataTest(unittest.TestCase):
+                )
+            )
+        with self.assertLogs(level=cf.log_level().value) as catch:
+            self.assertFalse(h.equals(i))
+            self.assertTrue(
+                any(
+                    "Data: Different array values" in log_msg
+                    for log_msg in catch.output
+                )
+            )
+
+        # Test masked arrays
+        j = cf.Data(np.ma.masked_all(shape, dtype="int"), "m")
+        with self.assertLogs(level=cf.log_level().value) as catch:
+            self.assertFalse(j.equals(d))
+            self.assertTrue(
+                any(
+                    "Data: Different array values" in log_msg
+                    for log_msg in catch.output
+                )
+            )
+
+        # Test rtol and atol parameters
+        k1 = cf.Data(np.array([10.0, 20.0]))
+        k2 = cf.Data(np.array([10.01, 20.01]))
+        for ks in [(k1, k2), (k2, k1)]:  # to test symmetry/commutativity
+            k_1, k_2 = ks
+            # Only one log check is sufficient here
+            with self.assertLogs(level=cf.log_level().value) as catch:
+                self.assertFalse(k_1.equals(k_2, atol=0.005, rtol=0))
+                self.assertTrue(
+                    any(
+                        "Data: Different array values (atol=0.005, rtol=0)"
+                        in log_msg
+                        for log_msg in catch.output
+                    )
+                )
+            self.assertTrue(k_1.equals(k_2, atol=0.02, rtol=0))
+            self.assertFalse(k_1.equals(k_2, atol=0, rtol=0.0005))
+            self.assertTrue(k_1.equals(k_2, atol=0, rtol=0.002))
+
+        # Test ignore_fill_value parameter
+        m1 = cf.Data(1, fill_value=1000)
+        m2 = cf.Data(1, fill_value=2000)
+        with self.assertLogs(level=cf.log_level().value) as catch:
+            self.assertFalse(m1.equals(m2))
+            self.assertTrue(
+                any(
+                    "Data: Different fill value: 1000 != 2000" in log_msg
+                    for log_msg in catch.output
+                )
+            )
+            self.assertTrue(m1.equals(m2, ignore_fill_value=True))
+
+        # Test verbose parameter: 1/'INFO' level is behaviour change boundary
+        for checks in [(1, False), (2, True)]:
+            verbosity_level, expect_to_see_msg = checks
+            with self.assertLogs(level=cf.log_level().value) as catch:
+                self.assertFalse(d2.equals(d, verbose=verbosity_level))
+                self.assertIs(
+                    any(
+                        "Data: Different data types: float32 != int64"
+                        in log_msg
+                        for log_msg in catch.output
+                    ),
+                    expect_to_see_msg,
+                )
+
+        # Test ignore_data_type parameter
+        # TODODASK - this one needs documenting in the method docstring.
+        # Question for DH: is ignore_type an alias?
```

Comment:
Hi @davidhassell, when you review this,  it would be good to get your advice regarding deprecated and potentially-deprecated arguments and how they should be managed, perhaps in a more general sense because a lot of them will apply across multiple methods.

Specifically for the `equals` method, the following are not documented:

* `ignore_data_type`;
* `ignore_type`;
*`ignore_compression`;

and I am not sure of the role of `ignore_compression` is not used in the method, suggesting it can be dropped?

Also, generally for this probable-`4.0.0` version, can we drop (some or all) long-deprecated methods like `traceback` completely or do we still want to handle these with the `_DEPRECATION*` functions, etc.?

Reply:
Hi Sadie, 

The undocumented keywords can easily be documented with docstring rewriter keys defined in cfdm, as is done in  `cfdm.Data.equals`.

`ignore_compression` is interesting, because it _is_ used in `cfdm.Data.equals` (which is a super method of `cf.Data,.equals`) so we need to review that separately 

I think you're right - at 4.0.0 we can drop all of the keywords that were deprecated at 3.0.0, and possibly some that were deprecated at later 3.x versions.

Thanks,
David

Reply:
Thanks for clarifying, @davidhassell, and for your review.

> The undocumented keywords can easily be documented with docstring rewriter keys defined in cfdm, as is done in cfdm.Data.equals.

Noted, thanks. I'll go in and do that where appropriate going forwards.

> ignore_compression is interesting, because it is used in cfdm.Data.equals (which is a super method of cf.Data,.equals) so we need to review that separately

Sure, I'll put in a specific `TODODASK` for that.

> I think you're right - at 4.0.0 we can drop all of the keywords that were deprecated at 3.0.0

That makes sense.

> and possibly some that were deprecated at later 3.x versions.

 I'll initiate a list covering such keywords in the main #182 thread, and add to it as I encounter them. We can then discuss them when there's a good time to to agree which to fully deprecate.